### PR TITLE
feat(http): introduce onError option on ServerInit

### LINKED ISSUE #1618: http: no error logged when calling undefined function
https://dash.deno.com/playground/whats-the-error

image

### LINKED ISSUE #1629: Errors thrown inside http.serve magically disappear
**Describe the bug**
Errors thrown inside http.serve magically disappear.

https://gist.github.com/Vehmloewff/775e44186cb2732523f1b0ecafbc9788

The error is never written to the console.

**Steps to Reproduce**

Run the example...

```sh
deno run --unstable --allow-net https://gist.githubusercontent.com/Vehmloewff/775e44186cb2732523f1b0ecafbc9788/raw/44d92c2a9e0d60923aa9461024fd715478787e12/errors-disappear.ts
```

...then head to http://localhost:3000

**Expected behavior**

"before throw" to be written to the console, then the error to be printed

**Environment**

- OS: Zorin OS 15.3
- deno version: 1.16.2
- std version: 1.113.0

### PR DESCRIPTION
**Pull request by satyarohith**
Closes #1618
closes #1629

### CODE REVIEW

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by AaronO** [comment]
Would be nice to provide an option to silence this, it will likely be undesirable noise for many production systems since real-world servers will get client-aborted connections semi-frequently

**Review by kt3k — APPROVED** [review]
LGTM

**Comment by kt3k** [comment]
This also closes #1629

**Comment by satyarohith** [comment]
> Would be nice to provide an option to silence this, it will likely be undesirable noise for many production systems since real-world servers will get client-aborted connections semi-frequently

I agree. Does `ignoreHandlerErrors` sound good? Or `suppressHandlerErrors` or `logHandlerErrors`?

[)](https://api.gh-polls.com/poll/01FN868S6THW2HKV3EF1AVHBQM/ignoreHandlerErrors%20(default%3A%20false)/vote)
[)](https://api.gh-polls.com/poll/01FN868S6THW2HKV3EF1AVHBQM/suppressHandlerErrors%20(default%3A%20false)/vote)
[)](https://api.gh-polls.com/poll/01FN868S6THW2HKV3EF1AVHBQM/logHandlerErrors%20(default%3A%20true)/vote)

I'll add the option to `ServerInit`.

**Comment by kt3k** [comment]
Another option might be `onError (default: (e) => console.log(e))`

**Comment by satyarohith** [comment]
> Another option might be `onError (default: (e) => console.log(e))`

I like this. It will provide more flexibility. On the name: should it be `onError` or `onHandlerError`? The latter sounds better to me as it's more explicit.

@AaronO thoughts?

**Review by lucacasonato — CHANGES_REQUESTED** [review]
Rather than logging all errors, could we log only the errors coming from the handler itself (not ones coming from `respondWith`)?

**Comment by ry** [comment]
@satyarohith `onError` is fine.

File: http/server.ts [hunk 1]
```diff
@@ -136,6 +144,20 @@ export class Server {
...
+        JSON.stringify(
+          {
+            error: "Internal Server Error",
+            message: error.message,
+          },
+          null,
+          2,
+        ),
+        { status: 500 },
+      );
```
Comment (ry) [inline http/server.ts:159]: Nice!

Probably should set the content type header to "application/json"

File: http/server.ts [hunk 2]
```diff
@@ -345,20 +367,23 @@ export class Server {
...
     try {
       // Handle the request event, generating a response.
-      const response = await this.#handler(
+      response = await this.#handler(
         requestEvent.request,
         connInfo,
       );
+    } catch (error) {
+      // Invoke onError handler when request handler throws.
+      response = await this.#onError(error as Error);
```
Comment (lucacasonato) [inline http/server.ts:379]: What should we do if this throws? :upside_down_face:

File: http/server.ts [hunk 3]
```diff
@@ -345,20 +367,23 @@ export class Server {
...
   ): Promise<void> {
+    let response: Response;
     try {
       // Handle the request event, generating a response.
-      const response = await this.#handler(
+      response = await this.#handler(
         requestEvent.request,
         connInfo,
       );
+    } catch (error) {
```
Comment (lucacasonato) [inline http/server.ts:377]:
```suggestion
    } catch (error: unknown) {
```

File: http/server.ts (hunk 2 above)
Comment (lucacasonato) [inline http/server.ts:379]:
```suggestion
      response = await this.#onError(error);
```

File: http/server.ts [hunk 4]
```diff
@@ -536,6 +561,9 @@ export interface ServeInit {
 
   /** An AbortSignal to close the server and all connections. */
   signal?: AbortSignal;
+
+  /** The handler to invoke when route handlers throw an error. */
+  onError?: (error: Error) => Response | Promise<Response>;
```
Comment (lucacasonato) [inline http/server.ts:566]:
```suggestion
  onError?: (error: unknown) => Response | Promise<Response>;
```

File: http/server.ts (hunk 2 above)
Comment (ry) [reply http/server.ts:379]: then the server should crash

**Review by lucacasonato — COMMENTED** [review]
Can we print just `Internal Server Error`, and not bother with this JSON stuff? Sending full error messages to the client could leak sensitive info and is probably not what the developer wants for production.

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by sant123** [comment]
This works for `serve` function only. Can this be implemented for both `serveTls` and `serveListener`? Just curious why it was not implemented for them.

**Comment by satyarohith** [comment]
> Can this be implemented for both `serveTls` and `serveListener`?

Yes. Would you like to make a PR? :)

> Just curious why it was not implemented for them.

I forgot to update them. :)

**Comment by sant123** [comment]
> 

Sure! Let me work on it 😉

**Comment by satyarohith** [comment]
@sant123, the issue has been resolved. Thanks!

**Pull request merged**